### PR TITLE
Revive-stacked-cards

### LINKED ISSUE #5385: Stacking / grouping items from same collection in explorer / bundles / spammy collection
In explorer when we are showing items, it's pretty useless for user to see **many same items** from one collection and it acts as visual spam.  "antispam card" 

I propose these filters in particular order
- using same IPFS hash - that classify definitely will be same visual item
- same price - indicates from one serial mint - if someone sets same price from minting, whatever.
- only one transfer (can still be some kind of airdrop from whitelist/allowlist)
- any other filters?

📟 @vikiival I guess we would need resolver for this kind of inspection on explorer to get reply faster from squid?

## Design
- we are going to replace multiple duplicated content with one card
image
- updated explore v0.9 figma: https://www.figma.com/file/OwAEyE0QDPn3tHXZbyot1o/Explore-v0.9?node-id=0%3A1&t=ezFOQUs4vhM0Txeu-1
image

### Basics 
- since we don't have a listings tab, this card will be only visible in explore/items
- card will show the name of the duplicate nft - when nft would be the same (img, description...) except name (Name #.23 and name #.45) is it possible to use this card as well - but without the number at the end? 
- how many times the nft is duplicated 
- floor price of all the same nfts
- collection name, the imgs are from 
- visit button 
- there is also small card:
  - used only on smaller previews 
  - absence of floor price, otherwise working similarly

### Functionality
- the nft card will work the same as other nft card
- stroke to the most contrast color in the mode + opacity overlay
- if you click the nft img, it will redirect you to the collection
- same with collection name and view 
- basically it's a shortcut to collection 

I would propose start doing this globally or at least on rmrk and rmrk2 as there is **massive amount of spam items** and we want to highlight good collections 

image

---
can test with 
https://beta.kodadot.xyz/rmrk2/u/DkjzQrt7U5WN66STqXmXeQYZkcd2mZuUqzncjXwDCck2wUQ?tab=collection

---
Lmk if you have any questions 🚀

### LINKED ISSUE #7462: Optimize query on TokenEntity
### What happened?

One of the reasons why #7419 on AHP was caused is this huge and scary query.
It does so many joins that squid database decided to give up. 

Screenshot 2023-10-02 at 12 59 10

```graphql
query tokenListWithSearch($first: Int!, $offset: Int, $denyList: [String!], $search: [NFTEntityWhereInput!], $orderBy: [TokenEntityOrderByInput!]) {
  tokenEntities: tokenEntities(
    orderBy: $orderBy
    limit: $first
    offset: $offset
    where: {nfts_some: {burned_eq: false, issuer_not_in: $denyList, metadata_not_eq: "", AND: $search}}
  ) {
    id
    blockNumber
    hash
    image
    media
    name
    updatedAt
    createdAt
    count
    nfts(
      where: {burned_eq: false, issuer_not_in: $denyList, metadata_not_eq: "", AND: $search}
    ) {
      ...nft
      ...nftDetails
      royalty
      recipient
      collection {
        id
        name
        floor
        __typename
      }
      meta {
        ...baseMeta
        __typename
      }
      __typename
    }
    __typename
  }
  tokenEntitiesConnection(
    where: {nfts_some: {burned_eq: false, issuer_not_in: $denyList, metadata_not_eq: "", AND: $search}}
    orderBy: blockNumber_DESC
  ) {
    totalCount
    __typename
  }
}

fragment nft on NFTEntity {
  id
  metadata
  issuer
  currentOwner
  blockNumber
  burned
  __typename
}

fragment nftDetails on NFTEntity {
  name
  sn
  price
  __typename
}

fragment baseMeta on MetadataEntity {
  id
  image
  animationUrl
  name
  description
  __typename
}

```

```json
{
	"search": [
		{}
	],
	"first": 40,
	"offset": 1,
	"denyList": [
		"128qRiVjxU3TuT37tg7AX99zwqfPtj2t4nDKUv9Dvi5wzxuF",
		"155HWw3J9jyYphMm5is4vp9Bzj7ZRRd6HEzCPdWd8cq97KfT"
	],
	"orderBy": [
		"blockNumber_DESC"
	]
}
```

### Please reproduce in steps

- #7419

### Expected Behavior

The task:

Delete unnecessary information from the query so I does not make too many complicated joins.

Blocked by Questions on @daiagi 

1. Do we really need to query all available NFTs?
2. Does all nfts need all metadata? 

### What browsers are you seeing the problem on?

_No response_

### At which address did you encounter bug?

_No response_

### Are you logged in?

None

### Which wallet you are using?

_No response_

### At which chain did you encounter bug?

_No response_

### Screenshots

photo_2023-10-02 12 59 21

### Relevant log output

_No response_

### Payment link for reward

https://kodadot.xyz/ksm/transfer/?target=%3CMy_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address%3E

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #7497: Incorrect floor for stacked cards
without any filters
image

with buy now filter: 
image

https://canary.kodadot.xyz/ahp/collection/11

### PR DESCRIPTION
**Pull request by daiagi**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [x] Feature
- [x] Refactoring

## Needs Design check

- @exezbcz please review

## Needs QA check

- @kodadot/qa-guild please review

## Context

- [x] Closes #7462
- [x] Closes #5385 (on AssestHub)
- [x] Closes #7497 (@exezbcz  see discussion below)
- [ ] Requires deployment <snek/rubick/worker>

#### Did your issue had any of the "$" label on it?

- [x] My DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)

## Screenshot 📸

- [x] My fix has changed UI

image

## More

@exezbcz
regarding #7497 :

the price shown is the actual lowest price of a stack - which is different from the collection floor price
that means the the price shown is the price user will pay for it
it also responds to filters  - meaning cout and price will change according to filter chosen

a UX issue (bug or feature i don't know) is that if ANY of the nfts under a stack has price = 0 (i.e. not listed for sell) => there is no buy now button until "buy now" filter is activated (or lowest price > 0 is set)

please DM me if any questions

### <samp>🤖 Generated by Copilot at f6f1b8d</samp>

This pull request adds support for token entities on the Asset Hub, which are NFTs with token ids that can be listed and bought as a group. It refactors several components, composables, and queries to handle both tokens and NFTs, and to display the correct price and metadata information for each type. It also introduces new types, props, and fields to represent the token entities and their associated NFTs.

### <samp>🤖 Generated by Copilot at f6f1b8d</samp>

> _Sing, O Muse, of the mighty deeds of the `ItemsGrid` component,_
> _How it displayed both tokens and NFTs with skill and judgment,_
> _Using the `useFetchSearch` function and the `ItemsItemsGridImageTokenEntity` subcomponent,_
> _And the `useNftActions` composable to handle the logic of buying and listing._

### CODE REVIEW

File: components/items/ItemsGrid/ItemsGridImage.vue [hunk 1]
```diff
@@ -135,7 +113,7 @@ const { cartIcon } = useShoppingCartIcon(props.nft.id)
 
 const { nft } = useNft(props.nft)
 
-const isOwner = computed(() =>
+const isUserTheOwner = computed(() =>
```
Comment (roiLeo) [inline components/items/ItemsGrid/ItemsGridImage.vue:116]: or `isCurrentOwner` could work out

**Comment by prury** [comment]
somehow related #7599

File: components/items/ItemsGrid/useNftActions.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
+import { isOwner as checkOwner } from '@/utils/account'
```
Comment (roiLeo) [inline components/items/ItemsGrid/useNftActions.ts:1]: move to composables?

File: components/items/ItemsGrid/useNftActions.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+export const userIsOwner = (entity: TokenEntity, accountId: string) =>
+  checkOwner(entity.cheapest?.currentOwner, accountId)
+
+export const isAvailableToBuy = (entity: TokenEntity) =>
+  Number(entity.cheapest.price) > 0
+
+const tokensNftsCache = new Map<string, NFTWithMetadata[]>()
+
+export const fetchNft = async (nftId: string): Promise<NFTWithMetadata> => {
```
Comment (roiLeo) [inline components/items/ItemsGrid/useNftActions.ts:18]: useAsyncData?

**Review by roiLeo — APPROVED** [review]

**Comment by prury** [comment]
loading super fast now ✔️ 
listing or adding to cart works fine ✔️ 
working properly on mobile ✔️ 
working fine on both ahk + ahp

liked that inside the collections don't show stacked no more

**Comment by yangwao** [comment]
Related to this pr? 

image

**Comment by prury** [comment]
> I can't see other items nor scroll down. Related to this pr?
> 
> image

here for me they load till page 23(end)

**Comment by exezbcz** [comment]
> the price shown is the actual lowest price of a stack - which is different from the collection floor price
> that means the the price shown is the price user will pay for it

how is that possible? Is that because of royalties or? I thought floor price is the lowest price. Ah maybe when there are more items that are not stacked right?

**Comment by exezbcz** [comment]
The small cards view has a different card height for stacked than the normal one: 
image

**Comment by daiagi** [comment]
@exezbcz

Take this collection for example

https://canary.kodadot.xyz/ahk/collection/6


It has the corn waifu nfts which are a single stack
And then there are other nfts in the collection


If one of these other nfts have price lower then corn waifu then you have a case where collection floor price is outside of the stack

Therefore displaying this lower price on the stack is irrelevant information


The collection floor price becomes recent again in the listing modal

**Comment by daiagi** [comment]
>The small cards view has a different card height for stacked than the normal one:

fixed

File: components/items/ItemsGrid/useNftActions.ts (hunk 3 above)
Comment (daiagi) [reply components/items/ItemsGrid/useNftActions.ts:18]: ill check

File: components/items/ItemsGrid/ItemsGridImage.vue (hunk 1 above)
Comment (daiagi) [reply components/items/ItemsGrid/ItemsGridImage.vue:116]: sure. it is a better name. thanks

**Comment by daiagi** [comment]
@yangwao @vikiival 

is there still interest in implementing this feature on
ksm/rmrk?
bsx?

File: components/items/ItemsGrid/useNftActions.ts (hunk 3 above)
Comment (daiagi) [reply components/items/ItemsGrid/useNftActions.ts:18]: I think you are referring to `useAsyncData` caching
1. I didn't know - thanks
2. not good enough here: in `itemsGrid` there is a single call to get all of the nfts of the tokens that _might_ get listed later
but each token's nfts are cached individually

the result is when user click `list` on a token that has already been fetched it isn't fetched again

to achieve that using `useAsyncData` ill need to split the initial query up, each token to its own query

**Comment by exezbcz** [comment]
@daiagi got it sir.

I think less confusing would be a change to "lowest price" 
image

File: components/items/ItemsGrid/ItemsGridImage.vue (hunk 1 above)
Comment (daiagi) [reply components/items/ItemsGrid/ItemsGridImage.vue:116]: * #7653

File: components/items/ItemsGrid/useNftActions.ts (hunk 2 above)
Comment (daiagi) [reply components/items/ItemsGrid/useNftActions.ts:1]: * #7653

**Comment by daiagi** [comment]
@exezbcz 
'lowest price' done

image

**Comment by daiagi** [comment]
@exezbcz 

from #7652 
>the current button is not displayed in the center of the card


fixed in this PR

File: utils/constants.ts [hunk 4]
```diff
@@ -55,8 +55,8 @@ export const URLS = {
     snek: 'https://squid.subsquid.io/snekk/graphql',
     snekRococo: 'https://squid.subsquid.io/sneck/graphql',
     marck: 'https://squid.subsquid.io/marck/graphql',
-    stick: 'https://squid.subsquid.io/stick/graphql',
-    speck: 'https://squid.subsquid.io/speck/graphql',
+    stick: 'https://squid.subsquid.io/stick/v/v5/graphql',
+    speck: 'https://squid.subsquid.io/speck/v/v5/graphql',
```
Comment (Jarsen136) [inline utils/constants.ts:59]: It should not be a fixed version.

Comment (daiagi) [reply utils/constants.ts:59]: ok then what to do?

it isn't deployed on stick/graphql

Comment (vikiival) [reply utils/constants.ts:59]: > when we release this feature

Can do a release if everything is good

cc @prury

**Comment by vikiival** [comment]
> is there still interest in implementing this feature on

ksm/rmrk?

- No

bsx?

- No

File: utils/constants.ts (hunk 4 above)
Comment (prury) [reply utils/constants.ts:59]: > > when we release this feature
> 
> Can do a release if everything is good
> 
> cc @prury

will check this one again in a few moments

**Comment by yangwao** [comment]
Meanwhile, can you @daiagi resolve conflicts?

@prury, how it stands?

**Comment by prury** [comment]
> Meanwhile, can you @daiagi resolve conflicts?
> 
> @prury, how it stands?

took a little while before images were loading properly, but now it seems ok

**Comment by yangwao** [comment]
And it seems branch needs an update? 

image

**Comment by prury** [comment]
> And it seems branch needs an update?
> 
> image

yup, this still needs solving

**Comment by daiagi** [comment]
>And it seems branch needs an update?

hmmm , tought I solved it already
will fix

**Comment by yangwao** [comment]
These gaps have been fixed in some PR no? 

image

**Comment by vikiival** [comment]
@daiagi I released the squid

Screenshot 2023-10-16 at 16 16 13

File: utils/constants.ts (hunk 4 above)
Comment (vikiival) [inline utils/constants.ts:59]:
```suggestion
    stick: 'https://squid.subsquid.io/stick/graphql',
    speck: 'https://squid.subsquid.io/speck/graphql',
```

**Comment by yangwao** [comment]
> hmmm , tought I solved it already
> will fix

@daiagi  you can do it in another PR as want to squeeze this one into the upcoming release

**Comment by yangwao** [comment]
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 3.77 USD/DOT ~ 13.263 $DOT
🧗 [1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8](https://polkadot.subscan.io/account/1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)
🔗 [0xd1d58b8281feda4d369e8bbe6f0b37c1d64ff942ac3b14118e6d7ed8d380b854](https://polkadot.subscan.io/extrinsic/0xd1d58b8281feda4d369e8bbe6f0b37c1d64ff942ac3b14118e6d7ed8d380b854)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**